src/SquidMath.h

```diff
@@ -117,11 +149,101 @@ NaturalSum(Args... args) {
 /// \returns the new variable value (like an assignment operator would)
 template <typename S, typename... Args>
 S
-SetToNaturalSumOrMax(S &var, Args... args)
+SetToNaturalSumOrMax(S &var, const Args... args)
 {
     var = NaturalSum<S>(args...).value_or(std::numeric_limits<S>::max());
     return var;
 }
 
+template <typename T>
+Optional<bool>
+NaturalValue(const T t)
+{
+    return t >= 0 ? Optional<bool>(bool(t)) : Optional<bool>();
```

Comment:
As we discussed, the `t>=0` comparison for unsigned types should produce a compiler warning. I made some tests to check this:

In **gcc**, the `-Wtype-limits` flag (or '-Wextra') is responsible for this warning:
```
warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
```

**clang** has `-Wtautological-unsigned-zero-compare` for this purpose:
```
warning: result of comparison of unsigned expression >= 0 is always true
```
This works as far as templates are not around. In the following example, the first (non-template) function produces the warning, while the template one does not:
```
bool
Foo1(unsigned long t) {
    return t>=0;
}

template <typename T>
bool
Foo2(const T t) {
    return t>=0;
}

unsigned long v = 5;
Foo1(v):
Foo2(v):

```

So (with `-Wtype-limits)` our testMath compiles without the warning. Maybe there are some additional flags controlling this behavior for templates, but I could not find them.

Reply:
Useful research, thank you. I wish `-Wtype-limits` would be documented as ignoring templated arguments, but I think we would have to live with a few GCC [bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=11856) [reports](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=82521) that mention that GCC should ignore such arguments (exactly for the reasons we would be struggling with these warnings in these Squid functions).

Reply:
However, we may not be able to get rid of Less<> completely because of `-Wsign-compare`:

```
../src/SquidMath.h:97:12: warning: comparison of integer expressions of different signedness:
‘const long unsigned int’ and ‘long int’ [-Wsign-compare]
        (t > std::numeric_limits<S>::max() - s) ? Optional<S>() :
tests/testMath.cc:264:5:   required from here
```

I continue to work on that...